### PR TITLE
[SPARK-17029] make toJSON not go through rdd form but operate on dataset always

### PR DESCRIPTION
**Pull request by robert3005**
## What changes were proposed in this pull request?

Don't convert toRdd when doing toJSON
## How was this patch tested?

Existing unit tests

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JsonFileFormat.scala [hunk 1]
```diff
@@ -155,13 +155,13 @@ class JsonFileFormat extends TextBasedFileFormat with DataSourceRegister {
 private[json] class JsonOutputWriter(
     path: String,
     bucketId: Option[Int],
-    dataSchema: StructType,
+    boundGen: (CharArrayWriter) => JacksonGenerator,
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JsonFileFormat.scala]: can you add classdoc to explain what boundGen (and other ctor parameters) does.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JsonFileFormat.scala [hunk 2]
```diff
@@ -84,7 +84,7 @@ class JsonFileFormat extends TextBasedFileFormat with DataSourceRegister {
           bucketId: Option[Int],
           dataSchema: StructType,
           context: TaskAttemptContext): OutputWriter = {
-        new JsonOutputWriter(path, bucketId, dataSchema, context)
+        new JsonOutputWriter(path, bucketId, new JacksonGenerator(dataSchema)(_), context)
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JsonFileFormat.scala]: what's the writer used here?

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 3]
```diff
@@ -2494,16 +2494,18 @@ class Dataset[T] private[sql](
    * @since 2.0.0
    */
   def toJSON: Dataset[String] = {
-    val rowSchema = this.schema
-    val rdd: RDD[String] = queryExecution.toRdd.mapPartitions { iter =>
+    import sparkSession.implicits.newStringEncoder
+    val rowSchema = schema
+    val boundGen = new JacksonGenerator(rowSchema)(_)
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: what's the writer used here?

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 4]
```diff
@@ -2494,16 +2494,18 @@ class Dataset[T] private[sql](
    * @since 2.0.0
    */
   def toJSON: Dataset[String] = {
-    val rowSchema = this.schema
-    val rdd: RDD[String] = queryExecution.toRdd.mapPartitions { iter =>
+    import sparkSession.implicits.newStringEncoder
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: where is the encoder used? I'd prefer to make this more explicit now it is farther apart from the import.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JsonFileFormat.scala (hunk 2 above)
Comment (robert3005) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JsonFileFormat.scala]: I thought it made more sense to apply the schema early since it's immutable over whole dataset while the writer is per partition. As you see in the JsonOutputWriter the writer is applied there. Not sure if this makes difference from perf point but it makes structure clearer (at least in my mind)

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala (hunk 4 above)
Comment (robert3005) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: The encoder is for the implicit mapPartitions argument. We need to have Encoder[String] for the output. Will put it just above the mappartitions call

**Comment by robert3005** [comment]
@rxin anything else? I added docs to the best of my understanding let me know if you meant something else.

**Comment by ash211** [comment]
This seems pretty reasonable, assuming test coverage already exists on that toJSON method.

Jenkins, this is ok to test.

**Comment by robert3005** [comment]
@rxin any chance you or someone else can take a look?

**Comment by ash211** [comment]
Jenkins, this is ok to test.

**Comment by ash211** [comment]
@robert3005 looks like this has unit test failures on `org.apache.spark.sql.hive.orc.OrcSourceSuite.SPARK-19459/SPARK-18220: read char/varchar column written by Hive` -- is that a flake?

**Comment by ash211** [comment]
Jenkins, this is ok to test.

**Comment by robert3005** [comment]
It indeed does look like a flake.

**Comment by robert3005** [comment]
ping? @rxin

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 5]
```diff
@@ -2692,9 +2692,10 @@ class Dataset[T] private[sql](
    * @since 2.0.0
    */
   def toJSON: Dataset[String] = {
-    val rowSchema = this.schema
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: Revert this change?

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 6]
```diff
@@ -2717,8 +2718,6 @@ class Dataset[T] private[sql](
         }
       }
     }
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: Instead of importing it, we can explicitly pass the encoder. 
```Scala
mapPartitions {
// func
} (sparkSession.implicits.newStringEncoder)
```

**Comment by robert3005** [comment]
thanks @gatorsmile, updated

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 7]
```diff
@@ -2828,9 +2828,7 @@ class Dataset[T] private[sql](
           json
         }
       }
-    }
-    import sparkSession.implicits.newStringEncoder
-    sparkSession.createDataset(rdd)
+    } (sparkSession.implicits.newStringEncoder)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: nit: `Encoders.STRING`

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala [hunk 8]
```diff
@@ -1326,6 +1327,17 @@ class JsonSuite extends QueryTest with SharedSQLContext with TestJsonData {
     )
   }
 
+  test("Dataset toJSON doesn't construct rdd") {
+    val containsRDD = spark.emptyDataFrame.toJSON.queryExecution.logical.find {
+      case ExternalRDD(_, _) => true
+      case _ => false
+    }
+
+    if (containsRDD.isDefined) {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala]: nit: `assert(containsRDD.isEmpty, "xxx")`

**Comment by cloud-fan** [comment]
LGTM

**Comment by cloud-fan** [comment]
thanks, merging to master!
